internal/rest/leveldbreceipts.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/kaleido-io/ethconnect/internal/kvstore"
+)
+
+type levelDBReceipts struct {
+	conf  *LevelDBReceiptStoreConf
+	store kvstore.KVStore
+}
+
+func newLevelDBReceipts(conf *LevelDBReceiptStoreConf) (*levelDBReceipts, error) {
+	store, err := kvstore.NewLDBKeyValueStore(conf.Path)
+	if err != nil {
+		return nil, errors.Errorf(errors.ReceiptStoreLevelDBConnect, err)
+	}
+	return &levelDBReceipts{
+		conf:  conf,
+		store: store,
+	}, nil
+}
+
+// AddReceipt processes an individual reply message, and contains all errors
+// To account for any transitory failures writing to mongoDB, it retries adding receipt with a backoff
+func (l *levelDBReceipts) AddReceipt(requestID string, receipt *map[string]interface{}) (err error) {
+	// insert an entry with a composite key to track the insertion order
+	timeElapsed := ((*receipt)["headers"].(map[string]interface{}))["timeElapsed"].(float64)
+	timeSeconds := (*receipt)["receivedAt"].(int64) + int64(timeElapsed)
+	timeStringSeconds := fmt.Sprintf("%d", timeSeconds)
+	timeStringNanoSeconds := fmt.Sprintf("%.9f", timeElapsed)
+	timeStringNanoSeconds = timeStringNanoSeconds[strings.Index(timeStringNanoSeconds, "."):]
+	timestamp := timeStringSeconds + timeStringNanoSeconds
+
+	// add "z" prefix so these entries come after the actual entries
+	// because for iteration we start from last backwards
+	compositeKey := fmt.Sprintf("z%s:%s", timestamp, requestID)
+	l.store.Put(compositeKey, nil)
+
+	// inser the actual entry
+	b, _ := json.MarshalIndent(receipt, "", "  ")
+	return l.store.Put(requestID, b)
+}
+
+// GetReceipts Returns recent receipts with skip & limit
+func (l *levelDBReceipts) GetReceipts(skip, limit int, ids []string, sinceEpochMS int64, from, to string) (*[]map[string]interface{}, error) {
+	itr := l.store.NewIterator()
+	defer itr.Release()
+	results := make([]map[string]interface{}, 0, limit)
+	index := 0
+	var errResult error
+	for valid := itr.Last(); valid; valid = itr.Prev() {
+		if limit > 0 && len(results) >= limit {
+			break
+		}
+
+		if index >= skip {
+			val := itr.Value()
+			if len(val) > 0 {
+				// the index keys don't have values. if we have hit an entry with a value
+				// then we have reached the end of the index keys
+				break
+			}
+			// extract the request key from the index key
+			r, _ := regexp.Compile("z[0-9.]+:(.+)")
+			compositeKey := itr.Key()
+			matches := r.FindStringSubmatch(compositeKey)
+			var lookupKey string
+			if len(matches) == 2 {
+				lookupKey = matches[1]
+			} else {
+				fmt.Printf("Failed to parse composite key %s\n", compositeKey)
+			}
+			val, err := l.store.Get(lookupKey)
+			if err != nil {
+				fmt.Printf("Failed to locate record by key %s\n", lookupKey)
+				continue
+			}
+			receipt := make(map[string]interface{})
+			err = json.Unmarshal(val, &receipt)
+			if err != nil {
+				fmt.Printf("Failed to decode stored receipt for request %s. This record is skipped in the returned list", itr.Key())
```

Comment:
Could we move these all over to `logrus` please.
There's no linter to protect against `fmt.Printf` at the moment, so it relies on a manual review to avoid these creeping in.

Reply:
(not commenting on all the items individually - but there are a number in the PR)